pkg/plugins/runtime/k8s/controllers/gateway_converter.go

```diff
@@ -0,0 +1,114 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	kube_core "k8s.io/api/core/v1"
+	kube_apierrs "k8s.io/apimachinery/pkg/api/errors"
+	kube_meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kube_types "k8s.io/apimachinery/pkg/types"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	kube_controllerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	mesh_k8s "github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/metadata"
+	k8s_util "github.com/kumahq/kuma/pkg/plugins/runtime/k8s/util"
+	util_proto "github.com/kumahq/kuma/pkg/util/proto"
+)
+
+// createorUpdateBuiltinGatewayDataplane manages the dataplane for a pod
+// belonging to a built-in Kuma gateway.
+func (r *PodReconciler) createorUpdateBuiltinGatewayDataplane(ctx context.Context, pod *kube_core.Pod) error {
+	dataplane := &mesh_k8s.Dataplane{
+		ObjectMeta: kube_meta.ObjectMeta{
+			Namespace: pod.Namespace,
+			Name:      pod.Name,
+		},
+		Mesh: k8s_util.MeshFor(pod),
+	}
+
+	gatewayName, _ := metadata.Annotations(pod.Annotations).GetString(metadata.KumaBuiltinGatewayAnnotation)
+	if gatewayName == "" {
+		return nil
+	}
+
+	dataplaneProto, err := r.PodConverter.BuiltinGatewayDataplane(pod, gatewayName)
+	if err != nil {
+		return errors.Wrap(err, "unable to translate a Gateway Pod into a Dataplane")
+	} else if dataplaneProto == nil {
+		// we don't want a dataplane, the existing object will be deleted
+		// through owner refs
+		return nil
+	}
+
+	spec, err := util_proto.ToMap(dataplaneProto)
+	if err != nil {
+		return err
+	}
+
+	operationResult, err := kube_controllerutil.CreateOrUpdate(ctx, r.Client, dataplane, func() error {
+		dataplane.Spec = spec
+
+		if err := kube_controllerutil.SetControllerReference(pod, dataplane, r.Scheme); err != nil {
+			return errors.Wrap(err, "unable to set Dataplane's controller reference to Pod")
+		}
+		return nil
+	})
+
+	if err != nil {
+		log := r.Log.WithValues("pod", kube_types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name})
+		log.Error(err, "unable to create/update Dataplane", "operationResult", operationResult)
+		r.EventRecorder.Eventf(pod, kube_core.EventTypeWarning, FailedToGenerateKumaDataplaneReason, "Failed to generate Kuma Dataplane: %s", err.Error())
+		return err
+	}
+
+	switch operationResult {
+	case kube_controllerutil.OperationResultCreated:
+		r.EventRecorder.Eventf(pod, kube_core.EventTypeNormal, CreatedKumaDataplaneReason, "Created Kuma Dataplane: %s", dataplane.Name)
+	case kube_controllerutil.OperationResultUpdated:
+		r.EventRecorder.Eventf(pod, kube_core.EventTypeNormal, UpdatedKumaDataplaneReason, "Updated Kuma Dataplane: %s", dataplane.Name)
+	}
+	return nil
+}
+
+func (p *PodConverter) BuiltinGatewayDataplane(
```

Comment:
It took me a while to recover my perspective on what is happening here, so we really need something that explains the workflow we are going for.

Since this is extracted from the Gateway API PR, the workflow is
1. Create a Gateway with a name
2. Create some kind of deployment, annotating the pods with name of the gateway.

This is almost the inverse of the Universal workflow, where you would create dataplanes with a set of tags and then create gateways to bind to them. I'm happy to discuss the pros and cons of this, but I think that consistency with the Universal workflow is a better place to start (since the Kubernetes controller can still build on that).

Reply:
My point of view is that this way is analogous to how we already do `Dataplane`s in Kubernetes, in that the user never has to deal with them (but does in Universal). I.e. this way is more consistent with the way we deal with Kubernetes vs Universal.

Reply:
> My point of view is that this way is analogous to how we already do `Dataplane`s in Kubernetes

Yeh I was thinking more along the lines of how ZoneIngress is handled.

> Additionally, we already handle the existing "gateway mode" of `Dataplane`s, which is at least somewhat comparable to the new builtin gateway, this way by [annotating a deployment as a gateway](https://kuma.io/docs/1.2.0/documentation/dps-and-data-model/#gateway).

Right. The `gateway` annotation would be great to use if it wasn't already taken.

I think that coupling Gateway and Dataplane at this later creates edge cases that we would otherwise not have to deal with. What happens when pods are created that reference a Gateway name that doesn't exist? What happens when the Gateway did exist, but was then deleted? And then the deployment needed to restart a pod?

In Universal, the workflow is
1. Create builtin gateway dataplanes with a service name
2. Create a Gateway with a service that will match the dataplanes you want

For this layer (the follow-on controller is different and OK to couple resources differently), I don't see the necessity of tying dataplanes to a specific Gateway resource. We need to know they will be a builtin gateway dataplane, and we need to know what tags to apply to them and that's really all.

I wonder whether it would work to extend the semantics of the gateway annotation so you could say `kuma.io/gateway: builtin`?

Reply:
> For this layer (the follow-on controller is different and OK to couple resources differently), I don't see the necessity of tying dataplanes to a specific Gateway resource. We need to know they will be a builtin gateway dataplane, and we need to know what tags to apply to them and that's really all.

I wonder whether it would work to extend the semantics of the gateway annotation so you could say kuma.io/gateway: builtin?

Makes sense. I've updated the PR to not depend on the Pod linking to a specific Gateway but rather to create the Dataplane using a tag based off the name of the Pod's Service and assume a Gateway exists that selects this same Dataplane (via the same tag).